### ISSUE TITLE
Screenshare broadcasts system-wide audio, not just that of the application. 

### ISSUE THREAD

**Issue by TheodorEvald** [Vencord/Vesktop#569]
**Describe the bug**

When streaming any application using Vesktop, audio from other applications is being picked up and broadcasted.

**Expected behavior**

Vesktop streaming only the audio of the application I selected to screenshare.

**Desktop (please complete the following information):**
 - OS/Distro: Windows 10 LTSC 21H2

**Additional Context**

I tried streaming VLC, Foobar2000, a variety of games and the issue persists.

**Comment by Vendicated** [comment]
yes this is intended

i don't think it's possible to stream just one window with electron api

**Comment by PolisanTheEasyNick** [comment]
Indeed, using electron API we can stream either all sound from Windows system or sound from one tab:
![image](https://github.com/Vencord/Vesktop/assets/39007846/9a57be62-6856-4bbb-ace5-c35fe7bbb192)
For now, we are setting `audio` parameter just to `loopback`:
https://github.com/Vencord/Vesktop/blob/main/src/main/screenShare.ts#L80

**Comment by Vendicated** [comment]
that api has no option to choose a custom window. WebFrameMain refers to an electron frame, aka a frame of your own app. not other apps

**Comment by PolisanTheEasyNick** [comment]
> that api has no option to choose a custom window. WebFrameMain refers to an electron frame, aka a frame of your own app. not other apps

Ohh, I see now. Anyway there's no way to choose another sound source :(

**Comment by Vendicated** [comment]
this issue does not apply to Linux. only Windows

**Comment by KIRUYXAN** [comment]
So it's not possible to fix this?

**Comment by Communified** [comment]
I love this program, but this being an "intended" feature is almost dealbreaking for the app; I can no longer stream in vc with audio without them being able hear all other sources of audio, INCLUDING themselves in the vc; which makes audio completely pointless. If there's truly no fix due to the api, it sounds like the api needs to be different for an issue as major as this; everything else functions as intended and how I'd expect within native discord. In discord, even desktop audio capture does NOT include discord; meaning you hear everything EXCEPT others voices, which is much more acceptable. From what you said does that mean it's impossible to fix even via a plugin strictly due to what the program is built off?

**Comment by mazeymoos0022** [comment]
Ope yeah I didnt see this one, didn't show up when I searched, sorry peeps, it's kinda annoying, and yes I wish there was a fix, but if ya can't fix it, that's all good. Hope there becomes a fix in the future! ❤

**Comment by quadrinex** [comment]
I maybe have a solution (for windows) tho it's not optimal but what if we set the output of a window to a virtual audio cable¹, loopback the audio¹ and use the virtual microphone for the stream. 

1.Can be done within windows settings

**Comment by CloveTwilight3** [comment]
> I maybe have a solution (for windows) tho it's not optimal but what if we set the output of a window to a virtual audio cable¹, loopback the audio¹ and use the virtual microphone for the stream.

I had a go looking at this, [commit 4db9f25](https://github.com/CloveTwilight3/Vesktop_Vencord/commit/4db9f2535d1ae630fadf7f2ed196d9c4e0b6b93b). Did you mean something like this?

**Comment by quadrinex** [comment]
> > I maybe have a solution (for windows) tho it's not optimal but what if we set the output of a window to a virtual audio cable¹, loopback the audio¹ and use the virtual microphone for the stream.
> 
> I had a go looking at this, [commit 4db9f25](https://github.com/CloveTwilight3/Vesktop_Vencord/commit/4db9f2535d1ae630fadf7f2ed196d9c4e0b6b93b). Did you mean something like this?

I'm not really good at coding but from what I understand that what I mean.

**Comment by bzarky** [comment]
yeah capturing the speech of the people you're streaming to is kind of a showstopper for any kind of streaming in windows vencord. 

guess i'm going back to vanilla discord

**Comment by rnxm** [comment]
how is it possible that the vanilla client doesn't share its own audio but vesktop does?

**Comment by SlyCedix** [comment]
If you want, you can manage this, though it's a bit involved.

On Windows OBS allows you to isolate individual program audio, as well as create a virtual camera.

On linux you need to install the `linux-pipewire-audio` plugin (to isolate audio) as well as the `v4l2loopback` kernel module and enable it with the `exclusive_caps=1` option (to create the virtual camera).

Once you've done that, you can just select the camera option in discord, select the virtual camera, and stream with whatever audio and video setup you want. Honestly it's pretty great to have that much control over the output of the stream, but it's definitely annoying to setup.

**Comment by JustArion** [comment]
@SlyCedix I think that the main issue currently is about maintainability. A solution has been proposed via a PR from @quadrinex but the code for it is a **bit much**. The code is primarily in C++ while most of Vesktop is written in TS / JS. It also doesn't help that @Vendicated looks to be primarily on Linux, making it harder or more annoying of an issue to test and maintain if / when bugs relating to the C++ code from the PR gets submitted.

While sure, the virtual camera workaround looks to work, it relies on third party programs when most people on Windows just expect the feature to work out of the box. I'm sure @quadrinex meant well with the PR, but I'm assuming its just a lot of a burden placed on the maintainers of Vesktop for such a PR to go through (again, given that the C++ code from the PR is quite a bit)

**Comment by SlyCedix** [comment]
Oh certainly, having everything work out of the box is ideal. I just had already found a bandaid for the problem so I wanted to share it. I was actually here to look at a different issue.

May 8, 2026 11:59:58 arion ***@***.***>:

**Comment by quadrinex** [comment]
I didn't find any way to make it lighter because electron doesn't support capturing an app audio natively, the solution I used is actually (I think) the same that obs uses to capture an app audio. I may search on making it lighter but i'm pretty sure its not possible to make it in TS / JS for now.

> [@SlyCedix](https://github.com/SlyCedix) A solution has been proposed via a PR from [@quadrinex](https://github.com/quadrinex) but the code for it is a **bit much**. The code is primarily in C++ while most of Vesktop is written in TS / JS.
> I'm sure [@quadrinex](https://github.com/quadrinex) meant well with the PR, but I'm assuming its just a lot of a burden placed on the maintainers of Vesktop for such a PR to go through (again, given that the C++ code from the PR is quite a bit)

**Comment by SlyCedix** [comment]
I wonder if using something like deno's FFI would allow implementing this feature in JS/TS. If that's built as a standalone app, then there is not much issue with running separate runtimes.

https://docs.deno.com/runtime/fundamentals/ffi/

May 8, 2026 17:38:30 Quadrinex ***@***.***>:

**Comment by Vendicated** [comment]
The issue isn't language or whatever. It's the fact that this adds additional complexity and something I will have to maintain (I don't even use Windows), for very little gain.

Screenshare on the official Discord app is great on Windows, Vesktop won't ever come close to it. Considering this fact, it'd be a wasted effort just to have an inferior alternative to an already existing solution. I consider this out of scope.

We provide the most basic Windows implementation for people who are okay with that. Anyone who wants something better should just use the official app for streaming.

**Comment by quadrinex** [comment]
Not really. FFI is fine for tiny WinAPI calls, but not for the actual WASAPI process-loopback capture.

That part needs async COM objects, callbacks, native audio buffers, event handles, and careful lifetime/thread handling. Doing it through JS FFI would be fragile and crash-prone, and you’d probably end up writing a native wrapper anyway.

> I wonder if using something like deno's FFI would allow implementing this feature in JS/TS. If that's built as a standalone app, then there is not much issue with running separate runtimes.
> 
> https://docs.deno.com/runtime/fundamentals/ffi/

---

Sadly, your correct, I should note that this implementation really isn't really stable and would add a lot of unnecessary complexity.

> The issue isn't language or whatever. It's the fact that this adds additional complexity and something I will have to maintain (I don't even use Windows), for very little gain.
> 
> Screenshare on the official Discord app is great on Windows, Vesktop won't ever come close to it. Considering this fact, it'd be a wasted effort just to have an inferior alternative to an already existing solution. I consider this out of scope.
> 
> We provide the most basic Windows implementation for people who are okay with that. Anyone who wants something better should just use the official app for streaming.

Now, will this issue be closed or will it remain open forever?

**Comment by SlyCedix** [comment]
Would a way to pick a camera as a source for the stream from a camera be a good compromise?

If someone from the community wants to make a nice easy companion app for it, that's great. If not, people can use the OBS workaround if they want.

Eventually, maybe a way to launch the application from vesktop later down the line, once it's matured.

May 8, 2026 18:38:09 V ***@***.***>:

**Comment by quadrinex** [comment]
For OBS you can right click on the canvas and click "Open Preview Projector" > "New Window"
Tho a good thing would be to change the audio source to a virtual mic

> Would a way to pick a camera as a source for the stream from a camera be a good compromise?
> 
> If someone from the community wants to make a nice easy companion app for it, that's great. If not, people can use the OBS workaround if they want.
> 
> Eventually, maybe a way to launch the application from vesktop later down the line, once it's matured.

**Comment by quadrinex** [comment]
@SlyCedix I made a pull request for using OBS virtual camera.

**Comment by SkyanUltra** [comment]
> The issue isn't language or whatever. It's the fact that this adds additional complexity and something I will have to maintain (I don't even use Windows), for very little gain.
> 
> Screenshare on the official Discord app is great on Windows, Vesktop won't ever come close to it. Considering this fact, it'd be a wasted effort just to have an inferior alternative to an already existing solution. I consider this out of scope.
> 
> We provide the most basic Windows implementation for people who are okay with that. Anyone who wants something better should just use the official app for streaming.

and yet, the majority of your user base who DOES use vesktop are pretty split between windows and linux users. people use this fork because of the security it provides, at least that's why i do. so why is maintaining an important aspect which as many other have cited as the primary reason that its driven them away from vesktop as a whole being treated as a problem of its own when it'd fix the biggest standing issue with the app currently?

just because official discord can manage it well doesn't mean you can't at least find SOME sort of solution, even if it isn't the best one. anything would be better than nothing, and considering how the readme for vesktop here is basically lying to windows users about being able to select audio sources (seriously, at least have a disclaimer about that or something) it just feels incredibly annoying. i don't want to compromise by using the official app because truthfully, i don't want to use bloatware filled with ads for random mobile game slop i'm never going to touch that scrapes my computer for any bit of data it can. i'd rather use the modded version of it which has security in place to protect your data better while i wait for stoat to update and reach the standard i need it to in order for me to comfortable switch to that.

**Comment by coreinsane** [comment]
@Vendicated, I’d really like to ask you to reconsider the native C++ approach proposed by @quadrinex, perhaps as an **experimental, opt-in Windows feature** rather than something that has to replace the current implementation.

For my use case there are actually two separate issues here, and they have very different priorities:

**Must-have:** people in the call must NOT hear themselves through my stream.

**Nice-to-have:** when streaming a specific application, only that application's audio should be captured instead of the entire system audio.

The first one is by far the important one for me. If my friends hear their own voices coming back through my stream, streaming with audio is effectively unusable. I can live with unrelated system sounds occasionally being included, but I cannot realistically use screensharing if everyone hears themselves with a delay.

I really appreciate the work @vocaldll did in #1294 to address exactly this problem with Chromium's `loopbackWithoutChrome`. However, the implementation currently only enables it on Windows builds >= 20348 and falls back to regular `loopback` below that.

I'm using Windows 10 22H2 (build 19045), so this unfortunately means that even once #1294 reaches a release, my system will still use regular loopback. Vesktop's own output, including my friends' voices, will still be captured by the stream.

This matters quite a lot to me because one of the main reasons I started using Vesktop was that it is such a compelling privacy-friendly alternative to the official Discord client. Vesktop itself even highlights that Discord gets much less access to the system, which is a major reason I'd prefer to stay with it.

That's why I think @quadrinex's native solution could still be extremely valuable even if maintaining a native C++ component isn't desirable as part of the default path.

Would it be possible to expose it behind something like:

**Experimental -> Native Windows audio capture**

with the current Chromium implementation remaining the default?

That seems like a good compromise:

* the normal/default implementation stays simple;
* the native component is clearly marked experimental and unsupported or best-effort;
* users explicitly opt into it;
* Windows 10 users like me would potentially have a solution for the self-echo problem;
* interested users could test it and provide feedback;
* proper per-application audio capture would be an additional benefit, rather than the main reason for having it.

For me personally, **excluding Vesktop/Discord's own audio is the important part**. Properly capturing only the selected application's audio would be great, but I would already be completely happy with an experimental solution that simply lets me stream system/application audio without my friends hearing themselves back.

I'd be very happy to enable it and test it on Windows 10 if such an experimental option were available.

**Comment by Vendicated** [comment]
Chromium fixed this themselves, you can try it out: https://github.com/Vencord/Vesktop/commit/cb9c55fe1aa4cd1d3fbe23bddfdca02627c8c372
